[조병민] #2-서울시OpenAPI 데이터 수집 / 가공출력 API 기능 개발

bmcho opened the pull request:
서울 하수관로 수위, 강우량 정보 OpenAPI 데이터를 받아오는 API 기능 구현

pekopekopekopayo commented on seoul/urls.py:
```diff
@@ -0,0 +1,5 @@
+
+from .views import SeoulSewerLevelRainFallView
+
+urlpatterns = [path("/gu-search/<gu_name>", SeoulSewerLevelRainFallView.as_view())]
```
<str: gu_name>형식으로는 안되나요?

bmcho replied:
가능합니다. datetype을 지정을 하지 않았을때에는 default값이 string이기에 두었었습니다.

pekopekopekopayo replied:
default가 str이라면 변경하지 않으셔도 됩니다.

pekopekopekopayo commented on labq/settings.py:
```diff
@@ -11,33 +11,34 @@
         return os.environ[var_name]
     except KeyError:
-        error_msg = f'Set the {var_name} environment variable'
+        error_msg = f"Set the {var_name} environment variable"
```
커밋내용과는 무관한 내용인것 같습니다.

bmcho replied:
이 내용은 기존 initialize 때부터 있는 코드입니다 싱글쿼터가 더블쿼터로 바뀐것 제가 지금 lint, autoformatting 설정을 해놓아서입니다

pekopekopekopayo replied:
그렇다면 변경사항에 추가시키지 않는편이 좋을 것 같습니다.

bmcho replied:
적용 했습니다.

pekopekopekopayo commented on labq/settings.py:
```diff
@@ -52,43 +53,44 @@ def get_env_variable(var_name):
+    "django.contrib.staticfiles",
+    "rest_framework",
+    "django_extensions",
+    "drf_yasg",
```
pip install drf_yasg을 안하면 에러가 발생합니다.
requirements.txt에 내용을 기입 부탁드립니다.

bmcho replied:
알겠습니다

bmcho replied:
적용했습니다.

pekopekopekopayo commented on seoul/views.py:
```diff
@@ -0,0 +1,126 @@
+        data = {
+            "gu_name": gu_name,
+            "avg_water_level": waterlevel_by_hour,
+            "sum_rain_fall": rianfall_by_datetime[0],
```
우량계별 측정이기 때문에 모든 우량계에 대한 데이터는 필요없을 것 같습니다.

bmcho replied:
제거 한다는 것을 빠뜨렸네요 :)

bmcho replied:
적용했습니다.

pekopekopekopayo commented on seoul/views.py:
```diff
@@ -0,0 +1,126 @@
+        rianfall_URL = f"http://openapi.seoul.go.kr:8088/{OPEN_API_KEY}/json/ListRainfallService/1/100/{gu_name}"
+        rianfall_response = json.loads(requests.get(rianfall_URL).content)
+
+        if "RESULT" in rianfall_response:
```
무슨역활을 하는 조건문인가요?

bmcho replied:
OpenAPI에서 Row data가 존재 하지 않을 시 최상위 카테고리 명을 RESULT로 반환합니다 
반환 할 데이터가 없다고 판단하기 위함입니다.

pekopekopekopayo replied:
![image](https://user-images.githubusercontent.com/91131029/176674946-2c9d9cc8-6b36-477c-800f-23dd51de4d81.png)
현재 작업브랜치를 작동시킨 결과입니다. row date가 있어도 result는 존재하고있습니다.

bmcho replied:
Result는 row data가 없을시에만 들어옵니다

pekopekopekopayo replied:
RESULT가 있었을때에 대한 조건문이 다른사람이 보았을때 애매할수도있으므로
```
if not("ListRainfallService" in rianfall_response):
```
과 같은형태로 수정 부탁드립니다.

bmcho replied:
적용했습니다

pekopekopekopayo commented on seoul/views.py:
```diff
@@ -0,0 +1,126 @@
+        waterlevel_URL = f"http://openAPI.seoul.go.kr:8088/{OPEN_API_KEY}/json/DrainpipeMonitoringInfo/1/1000/{gu_id}/{before_onehour_date}/{before_onehour_date}"
+        waterlevel_response = json.loads(requests.get(waterlevel_URL).content)
+
+        if "RESULT" in waterlevel_response:
```
위와 같습니다

bmcho replied:
위와 같습니다.

pekopekopekopayo replied:
위에 내용 참고 부탁드립니다

pekopekopekopayo replied:
RESULT가 있었을때에 대한 조건문이 다른사람이 보았을때 애매할수도있으므로
```
if not("DrainpipeMonitoringInfo" in rianfall_response):
```
과 같은형태로 수정 부탁드립니다.

bmcho replied:
적용했습니다

pekopekopekopayo reviewed (commented):
확인 부탁 드립니다.

bmcho committed:
#2-modify-seoul-view

doble-quote -> single-quote 변경

bmcho committed:
#2-modify-seoul-view

66,77번째 줄 row data 확인 코드 변경
RESULT는 다른 사람에게 있어 혼란을 줄 수 있어
조건을 '데이터 key가 없을 시'로 변경

bmcho committed:
#2-fix-settings

drf-yasg ->  drf_yasg  import 오류 변경

pekopekopekopayo commented on seoul/views.py:
```diff
@@ -0,0 +1,118 @@
+        RianFallData
+        """
+        rianfall_by_raingauge_dict = {
+            i: round(sum(map(lambda x: float(x['RAINFALL10']), j)), 2)
```
map for 을 두번식 사용하여 시간복잡도를 저해하고있습니다.
결과 값이 정확하지 않습니다.
map(lambda x: x[date] ? ~~~) 같은 형식으로 바꾸거나 메소드를 사용하여 코드 수정 부탁드립니다.

bmcho replied:
확인했습니다.

pekopekopekopayo commented on seoul/views.py:
```diff
@@ -0,0 +1,118 @@
+            i: round(sum(map(lambda x: float(x['RAINFALL10']), j)), 2)
+            for i, j in rianfall_groupby_raingauge
+            for k in j
+            if re.sub('[-:\s]', '', k['RECEIVE_TIME'])[:10] == before_onehour_date
```
if문의 사용방법이 틀리셨습니다.

bmcho replied:
확인했습니다,

bmcho committed:
#2-fix-seoul-view-rainfalldata

list-comprehension 코드 작성 에러로 결과 값과 맞지 않은 값이 출력 되고 있었음. 
날짜 비교를 lambda 안으로 넣음으로 해결

pekopekopekopayo commented on seoul/serializers.py:
```diff
@@ -0,0 +1,21 @@
+
+class SeoulSewerLevel(serializers.Serializer):
+    IDN = serializers.IntegerField()
+    GUBN = serializers.CharField(max_length=10)
```
유효성 검사에 필요성이 있나요?

pekopekopekopayo commented on seoul/serializers.py:
```diff
@@ -0,0 +1,21 @@
+class SeoulSewerLevel(serializers.Serializer):
+    IDN = serializers.IntegerField()
+    GUBN = serializers.CharField(max_length=10)
+    GUBN_NAM = serializers.CharField(max_length=50)
```
위와같습니다.

pekopekopekopayo commented on seoul/serializers.py:
```diff
@@ -0,0 +1,21 @@
+
+
+class SeoulRainGauge(serializers.Serializer):
+    raingauge_name = serializers.CharField(max_length=50)
```
위와같습니다.

pekopekopekopayo commented on seoul/serializers.py:
```diff
@@ -0,0 +1,21 @@
+
+
+class SeoulSewerLevelRainFall(serializers.Serializer):
+    gu_name = serializers.CharField(max_length=20)
```
위와같습니다.

pekopekopekopayo commented on seoul/views.py:
```diff
@@ -0,0 +1,123 @@
+        }
+        serializer = SeoulSewerLevelRainFall(data=data)
+
+        if not serializer.is_valid():
```
유효성 검사에 대한 필요성이 있나요?

bmcho replied:
serializer는 데이터 검증이 필수 사항입니다 :) 안 쓰면 에러나요.

heyhyo11 commented on seoul/views.py:
```diff
@@ -0,0 +1,120 @@
+        before_onehour_date = (now - timedelta(hours=1)).strftime('%Y%m%d%H')
+
+        """
+        RianFall GET API
```
class 이름은 rainfall인데 변수명은 rianfall 입니다. 통일되면 좋을 것 같습니다.

bmcho replied:
확인했어요 :)

bmcho committed:
#2-modify-Api seoul typo And Unnecessary

불필요한 코드 제거 및 코드 오타 수정
